Sources/ComposableArchitecture/Instrumentation.swift

```diff
@@ -0,0 +1,142 @@
+///   11. The ``ViewStore`` is updated with the ``Store.state``s new value
+///   12. The ``Instrumentation.ViewStore.willDeduplicate`` callback is called
+///   13. The state is deduplicated
+///   14. The ``Instrumentation.ViewStore.didDeduplicate`` callback is called
+///   15. The ``Instrumentation.ViewStore.stateWillChange`` callback is called
+///   16. The ``ViewStore.state`` value is updated
+///   17. All subscribers to the ``ViewStore.state`` are updated with the new state value
+///   18. The ``Instrumentation.ViewStore.stateDidChange`` callback is called
+///   19. The ``Instrumentation.Store.didScope`` callback is called
+///   20. The ``Instrumentation.Store.didSend`` callback is called
+///   21. The ``Instrumentation.ViewStore.didSend`` callback is called
+public class Instrumentation {
+  public typealias Trigger = (EventInfo) -> Void
+
+  /// Container for the information that will be provided to tracking/instrumentation implementations.
+  public struct EventInfo: CustomStringConvertible {
+    internal init(type: String, action: String = "", tags: [String: String] = [:]) {
+      self.type = type
+      self.action = action
+      self.tags = tags
+    }
+
+    /// The Swift type of object that is operating. This will generally be of the type `Store<State, Action>` or
+    /// `ViewStore<State, Action>` with the `State` and `Action` types properly filled in. With this information it
+    /// _should_ be possible to identify which ``Store`` or ``ViewStore`` (or other type) is operating.
+    public let type: String
+
+    /// A ``String`` generated for the ``Action`` that was sent, when available. There are operations that may not have
+    /// an ``Action`` available and so this may be an empty string. The value is generated using ``String(describing:)``
+    /// and so is dependent on the Swift runtime metadata. If that metadata is removed in some way then this value will
+    /// be empty most likely.
+    public var action: String
+
+    /// A dictionary of tags that the library thinks are valuable to include in the tracking/instrumentation. There is
+    /// obviously no requirement to use these, but they are there just in case.
+    public var tags: [String: String]
+
+    public var description: String {
+      guard !action.isEmpty else {
+        return "\(type)"
+      }
+
+      return "\(type): \(action)"
+    }
+  }
+
+  public init(viewStore: Instrumentation.ViewStore? = nil, store: Instrumentation.Store? = nil) {
+    self.viewStore = viewStore
+    self.store = store
+  }
+
+  public static var shared: Instrumentation = .noop
+
+  /// Tracking/instrumentation hooks that operate only within the context of ``ViewStore`` objects.
+  public struct ViewStore {
+    public init(willSend: @escaping Trigger, didSend: @escaping Trigger, willDeduplicate: @escaping Trigger, didDeduplicate: @escaping Trigger, stateWillChange: @escaping Trigger, stateDidChange: @escaping Trigger) {
```

Comment:
ISTM the Instrumentation API could be simplified significantly by moving the type of the event into EventInfo, and registering only a single Trigger.
(i.e. EventInfo.Type would be an enum with values like 
.ViewStoreWillSend
.ViewStoreDidSend
etc)

Could simplify further by moving the will/did split into a new field on EventInfo.